xds/internal/xdsclient/pool.go

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/internal/backoff"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/xds/bootstrap"
+)
+
+var (
+	// DefaultPool is the default pool for xds clients. It is created at the
+	// init time.
+	DefaultPool *Pool
+)
+
+// Pool represents pool of xds clients.
+type Pool struct {
+	mu      sync.Mutex
+	clients map[string]*clientRefCounted
+	config  *bootstrap.Config
+}
+
+func init() {
+	DefaultPool = &Pool{clients: make(map[string]*clientRefCounted)}
+}
+
+// NewPool creates a new xds client pool with the given bootstrap contents.
+func NewPool(bootstrapContents []byte) (*Pool, error) {
+	config, err := bootstrap.NewConfigForTesting(bootstrapContents)
+	if err != nil {
+		return nil, err
+	}
+	return &Pool{
+		clients: make(map[string]*clientRefCounted),
+		config:  config,
+	}, nil
+}
+
+// NewClient returns an xds client with the given name from the pool. If the
+// client doesn't already exist, it creates a new xds client and adds it to the
+// pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+func (p *Pool) NewClient(name string) (XDSClient, func(), error) {
+	config, err := bootstrap.GetConfiguration()
+	if err != nil {
+		return nil, nil, fmt.Errorf("xds: failed to get xDS bootstrap config: %v", err)
+	}
+	return newRefCounted(name, p, config, defaultWatchExpiryTimeout, defaultIdleChannelExpiryTimeout, backoff.DefaultExponential.Backoff)
+}
+
+// NewClientForTesting returns an xds client configured with the provided
+// options from the pool. If the client doesn't already exist, it creates a new
+// xds client and adds it to the pool.
+//
+// The second return value represents a close function which the caller is
+// expected to invoke once they are done using the client.  It is safe for the
+// caller to invoke this close function multiple times.
+//
+// # Testing Only
+//
+// This function should ONLY be used for testing purposes.
+func (p *Pool) NewClientForTesting(opts OptionsForTesting) (XDSClient, func(), error) {
+	if opts.Name == "" {
+		return nil, nil, fmt.Errorf("opts.Name field must be non-empty")
+	}
+	if opts.WatchExpiryTimeout == 0 {
+		opts.WatchExpiryTimeout = defaultWatchExpiryTimeout
+	}
+	if opts.IdleChannelExpiryTimeout == 0 {
+		opts.IdleChannelExpiryTimeout = defaultIdleChannelExpiryTimeout
+	}
+	if opts.StreamBackoffAfterFailure == nil {
+		opts.StreamBackoffAfterFailure = defaultStreamBackoffFunc
+	}
+	if opts.Contents != nil {
+		config, err := bootstrap.NewConfigForTesting(opts.Contents)
+		if err != nil {
+			p.mu.Unlock()
```

Comment:
For my understanding, where is this `p.mu.Lock()` called?

Reply:
yeah good catch. That was supposed to be removed. I removed it.